### PR TITLE
fix: HasEntrypoints issues #1385 and #1260

### PR DESCRIPTION
**Pull request by miparnisari**
## Description
Alternative PR to https://github.com/openfga/openfga/pull/1405.

This one only fixes the issues with no refactors.

## References
- Close https://github.com/openfga/openfga/issues/1385
- Close https://github.com/openfga/openfga/issues/1260

### CODE REVIEW

File: pkg/typesystem/typesystem.go [hunk 1]
```diff
@@ -709,7 +713,11 @@ func hasEntrypoints(
 				return false, false, fmt.Errorf("undefined type definition for '%s#%s'", assignableTypeName, assignableRelationName)
 			}
 
-			if _, ok := v[assignableTypeName][assignableRelationName]; ok {
+			if hasEntrypoint, ok := v[assignableTypeName][assignableRelationName]; ok {
+				if hasEntrypoint {
+					return true, false, nil
```
Comment (miparnisari) [inline pkg/typesystem/typesystem.go:718]: i think these lines should stay (to be consistent with line 768), but i can't think of a test that covers this case :(

File: pkg/typesystem/typesystem.go [hunk 2]
```diff
@@ -620,11 +624,15 @@ func hasEntrypoints(
...
+		// At least one type must have an entrypoint.
 		for _, assignableType := range tuplesetRelation.GetTypeInfo().GetDirectlyRelatedUserTypes() {
 			assignableTypeName := assignableType.GetType()
 
 			if assignableRelation, ok := typedefs[assignableTypeName][computedRelationName]; ok {
-				if _, ok := v[assignableTypeName][computedRelationName]; ok {
+				if hasEntrypoint, ok := v[assignableTypeName][computedRelationName]; ok {
+					if hasEntrypoint {
+						return true, false, nil
+					}
```
Comment (miparnisari) [inline pkg/typesystem/typesystem.go:635]: for #1260

File: pkg/typesystem/typesystem.go [hunk 3]
```diff
@@ -601,8 +605,8 @@ func hasEntrypoints(
 			return false, false, fmt.Errorf("undefined type definition for '%s#%s'", typeName, computedRelationName)
 		}
 
-		if _, ok := v[typeName][computedRelationName]; ok {
-			return false, true, nil
+		if hasEntrypoint, ok := v[typeName][computedRelationName]; ok {
+			return hasEntrypoint, true, nil
```
Comment (miparnisari) [inline pkg/typesystem/typesystem.go:609]: for #1385

File: pkg/typesystem/typesystem.go [hunk 4]
```diff
@@ -565,8 +567,10 @@ func hasEntrypoints(
 
 	switch rw := rewrite.GetUserset().(type) {
 	case *openfgav1.Userset_This:
+		// At least one type must have an entrypoint.
 		for _, assignableType := range relation.GetTypeInfo().GetDirectlyRelatedUserTypes() {
 			if assignableType.GetRelationOrWildcard() == nil || assignableType.GetWildcard() != nil {
+				v[typeName][relationName] = true
```
Comment (miparnisari) [inline pkg/typesystem/typesystem.go:573]: for both #1260 and #1385

**Comment by jon-whit** [comment]
Nice changes! Looks solid 👍 

We'll want to revisit the case, for example:
```
type user
type employee

type document
    relations
        define editor: [employee]
        define viewer: [user] and editor  
```
and
```
type user
type employee

type document
  relations
    define restricted: [employee]
    define viewer: [user] but not restricted
```

The question is whether or not these should have entrypoints? In terms of semantics, should a relation involving intersection or exclusion have non-zero entrypoints if the two operands do not share at least one overlapping terminal subject type? My intuition tells me that the models above should have 0 entrypoints.

**Review by jon-whit — APPROVED** [review]

**Review by jon-whit — APPROVED** [review]

**Review by willvedd — APPROVED** [review]

**Pull request merged**